code-server: init at 3.4.1

offlinehacker opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

Package `code-server` and build it from source.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

offlinehacker committed:
code-server: init at 3.4.1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/vscode-remote-development-on-nixos/3119/36

offlinehacker commented:
@GrahamcOfBorg build code-server

offlinehacker commented:
@GrahamcOfBorg build code-server

offlinehacker commented:
@GrahamcOfBorg build code-server

offlinehacker commented:
@GrahamcOfBorg build code-server

offlinehacker commented:
@GrahamcOfBorg build code-server

offlinehacker commented:
@GrahamcOfBorg build code-server

offlinehacker commented:
I am unsure why it fails on aarch64-linux, i will remove it from supported platforms for now

offlinehacker commented:
@GrahamcOfBorg build code-server

Ma27 commented on pkgs/servers/code-server/default.nix:
```diff
@@ -0,0 +1,139 @@
+    # to get hash values use nix-build -A code-server.yarnPrefetchCache --argstr system <system>
+    outputHash = {
+      x86_64-linux = "197v9660yqhvazalyhc87ciy6q2ii8xxi5bc5y96bck2w0d2y8if";
+      aarch64-linux = "LiIvGuBismWSL2yV2DuKUWDjIzuIQU/VVxtiD4xJ+6Q=";
```
I'm not sure if we want to add even more packages to `nixpkgs` utilizing this controversial pattern.

offlinehacker replied:
Explain me what do you mean by that? I think this is most clean way. What is your proposed alternative?

Ma27 replied:
The issue is described here: https://github.com/NixOS/nix/issues/2270

Although I know that `yarn2nix` can be painful in some cases, is there any reason not to use it here?

offlinehacker replied:
Two major issues with using generator tools are:

- large generated files (maybe ~10mb in case of code-server), which I cannot put into nixpkgs repo as they just unnecessary pollute repo.
- community maintained tool, that tries to wrap other package manager, usually in imperfect way and can present much pain for the one creating or updating the package

I think static output hashes are best tool we have. Personally I would want to throw away purity for some derivations, as I trust underlying package manager, but nix philosophy is well to be pure, but this sometimes present great pain when just trying to package something, already spending maybe days figuring out how to do it "nix way".
Sometimes I just want to go back to Debian, as sometimes I do not care for purity, as reality is imperfect and unpure.

offlinehacker replied:
Also in many cases, less magic is better, less headache, more fun.

Ma27 replied:
Don't get me wrong: while I agree with the reasoning in the issue I linked above, I don't think we should prohibit this entirely, instead this should still be an option for software which is hard to package.

I mainly wanted to make sure that packagers are aware of this and using such a fixed-output drv is the best option :)

offlinehacker replied:
Ok, I understand, I just don't want to make packaging too hard and confusing due to multiple layers of abstractions, for a goal that's not even that important in some cases. I had quite some fight with lang2nix tools (implementing pypi2nix long time ago and some node2nix stuff) and when I look back, it was really not worth it for 99% cases.

Regarding this package if there would be a way to store large generated files in some other storage, I could use yarn2nix, but well at the end is the same, just one is more simple. One large file with a lot of hashes or one hash that represents them all.

offlinehacker replied:
What am I really missing is better tooling for static output derivations, as for example this is what I find ugly: https://github.com/NixOS/nixpkgs/pull/87258/files#diff-97ddd5942a260ac035c022c0c57de234R122

offlinehacker replied:
Oh yeah, recursive nix would help here, since i could use yarn2nix as part of build process and have generation process run as part of it.

As soon as recursive nix is supported on hydra, i can rewrite it. Still we would only have one static output hash, but process of creating deterministic output would be more stable, cached and arch independent. This i think is also a solution to improve static output derivations.

Mic92 replied:
Can be recursive nix be used in this case even? I don't think the inner nix instance has access to network or can create its own derivations as far as I know.

offlinehacker replied:
Regarding network I do not know, but it should be able to create derivations, since isn't that the point of recursive nix? But yes under assumption recursive nix can access network running under static output derivation, this should be possible.

Mic92 replied:
The intention of recursive nix was to abuse nix as a build system like here: https://github.com/edolstra/nix-make without paying having to setup a sandbox for each build. I assume it can access the network if the outer nix build has it, but I don't think it can have its own derivation.

offlinehacker commented:
I am using this package for some time, without any issues. Are we ready to merge it?

Mic92 commented:
Works on my machine (tm). There is apparently also an upgrade available.

Mic92 commented:
Maybe also patch out references to `/sbin/`

```
1warn  vscode Unable to retrieve mac address (Error: Command failed: /sbin/ifconfig -a || /sbin/ip link
/bin/sh: /sbin/ifconfig: No such file or directory
/bin/sh: /sbin/ip: No such file or directory
```

FRidh commented on pkgs/servers/code-server/default.nix:
```diff
@@ -0,0 +1,139 @@
+    mv node-* $out
+  '';
+
+  nativeBuildInputs = [ nodejs yarn python pkgconfig zip makeWrapper git rsync ];
```
surely not Python 2 is needed?

lilyball replied:
The [docs on building vscode](https://github.com/Microsoft/vscode/wiki/How-to-Contribute#build-and-run) say

> * [Python](https://www.python.org/downloads/release/python-2715/), at least version 2.7 (version 3 is _**not**_ supported)
>   * **Note:** Python 2.7 will be automatically installed for Windows users through installing `windows-build-tools` npm module (see below)

Curiously the Homebrew formula for code-server declares a dependency on python3.8. I think someone was confused though based on the above, and it probably only works because macOS still ships with python 2 (though it likely won’t in macOS 10.16).

offlinehacker replied:
Explicitly set python to `python3`

lilyball replied:
I haven't tested this derivation myself, but since the vscode docs explicitly state that python 3 is not supported, don't we need to be using python 2 here?

I also find it confusing that the selection of python 3 is done in the `callPackage` instead of in the arguments to this derivation.

offlinehacker replied:
Ok, I can change to python2 explicitly and try to rebuild.

Is putting it in attribute set preffered way? The issue is when you want to override for example:

```
code-server.override { python2 = python3; }
```

instead of:

```
code-server.override { python = python3; }
```

The first one feels wierd, so that's my issue with putting in package itself.

offlinehacker replied:
But it looks like it builds with python3 just fine, why do we need to use python2 for building then?

offlinehacker replied:
I decided to let it build with python3, as this is only build time dependency, so there should be no issues in runtime, and as long as it builds fine and works fine it's fine for me. Idk what issues vscode has with python3 or if documentation is up to date, but it works fine for us. If there are any issues in the future we can still switch it to python2.

offlinehacker replied:
@lilyball i implemented suggested changes, except change to `python2`.

lilyball commented on pkgs/servers/code-server/default.nix:
```diff
@@ -0,0 +1,139 @@
+
+  passthru = {
+    prefetchYarnCache = overrideDerivation yarnCache (d: {
+      outputHash = "0000000000000000000000000000000000000000000000000000000000000000";
```
BTW `stdenv.lib.fakeSha256` exists with this same value, if you want to avoid embedding a long string of zeroes and maybe make it more semantically meaningful.

```suggestion
      outputHash = stdenv.lib.fakeSha256;
```

offlinehacker replied:
Done

lilyball commented on pkgs/servers/code-server/default.nix:
```diff
@@ -0,0 +1,139 @@
+, rsync, makeDesktopItem, runtimeShell, writeScript, runCommand
+, nodejs, yarn, libsecret, xorg, ripgrep, electron }:
+
+with stdenv.lib;
```
Is this even being used outside of the `meta` attribute? It’s a lot more common to use `meta = with stdenv.lib; { … }` than to import it like this at the top level.

If you do need easy access to `lib` it’s also more common to do something like `let inherit (stdenv) lib; in` or just put `lib` in the arguments.

offlinehacker replied:
Should be fixed now

lilyball commented:
code-server 3.4.0 is out now.

BTW when updating to 3.4.0, the `lib/vscode/build/builtInExtensions.json` file is gone and the data was migrated into `lib/vscode/product.json`. You’ll likely need to use something like `jq` to modify the file as needed, if you don’t want to maintain a patch.

breakds commented:
Thanks for this!

I tried it on my current 20.03 NixOS box, the client-side gets disconnected and asks for reconnection every 5 seconds or so. Anyone ran into the same problem?

wucke13 commented:
This works fine for me, @breakds .
From my point of with this LGTM.

breakds commented:
> This works fine for me, @breakds .
> From my point of with this LGTM.

Okay I figured out what went wrong. I did not use the nodejs 12 version. It worked after I explicitly specify that. Thank you!

offlinehacker commented:
I did following changes:

- Update to latest 3.4.1 version
- Explicitly set `nodejs` version to 12 and `python` version to 3

Missing changes (planning to implement):

- Disable notifications for automatic updates.

Please review, it can be merged without disabled automatic updates.

Mic92 commented:
Not a blocker, but do you know what this `iproute2` calls do? https://github.com/NixOS/nixpkgs/pull/87258#issuecomment-634998719

offlinehacker commented:
@Mic92 I have to check and possibly also fix that yes

offlinehacker commented:
@GrahamcOfBorg build code-server

offlinehacker commented:
Well it seems there is some regression if building from master (I did not rebase when updating), so I need to rebase and check why it fails.

offlinehacker commented:
Ahh I see that I forgot to patch shebangs while fetching yarn cache. For some reason it worked on my machine, as `/bin/sh` exists for me and it apparently is not isolated if running in static output hash derivation, wierd.

lilyball commented:
`/bin/sh` exists because POSIX requires it.

On Linux in the sandbox `/bin/sh` is mapped to a minimal busybox shell (this is done as a configure argument defined in the `nixpkgs.nix` derivation). On macOS the sandbox explicitly allows access to system `/bin/sh` since it doesn't support remapping paths (system `/bin/sh` here is Bash 3.2.57 running in sh compatibility mode).

offlinehacker commented:
@lilyball It does not seem to exist, when ofborg is building it: https://github.com/NixOS/nixpkgs/pull/87258/checks?check_run_id=750458903 which I think I right behavior.

offlinehacker commented:
@Mic92 I fixed path to ifconfig. vscode tries to get id of machine and it uses mac address (hashed) to do so, see here: https://github.com/microsoft/vscode/blob/c6c6969c3c1d5a975ed213d58a586ec1f1007c33/src/vs/base/node/id.ts#L95

Path is fixed now, so this will work.

offlinehacker commented:
@GrahamcOfBorg build code-server

offlinehacker commented:
@GrahamcOfBorg build code-server

offlinehacker commented:
Ok, I think this is now ready to be merged, it starts without any issues, does not bother with updates and looks good in general. I will deploy it on my dev environments and do some more testing, so i can confirm this version works ok.

wucke13 commented:
One small remark: The current state of this works for me, but `git`. Instead of the browser displaying `code-server`, the terminal in which `code-server` is started receives the dialog for my ssh key. Is this intended?

kamaln7 commented:
I appreciate your work on this! I just set it up on my machine and it works perfectly 👍  Would be happy to see it merged 😄 

I ended up installing it like this:

```nix
{ config, pkgs, ... }:
let
  xtruder = import
    (builtins.fetchTarball https://github.com/xtruder/nixpkgs/tarball/88ce5a75cbaaaf1830006501c85a6fba0c0d05c2)
    { config = config.nixpkgs.config; };
in {
  environment.systemPackages = with pkgs; [
    xtruder.code-server
  ];

  systemd.services.code-server = {
    enable = true;
    description = "Remote VSCode Server";
    after = ["network.target"];
    wantedBy = ["multi-user.target"];

    serviceConfig = {
      Type = "simple";
      ExecStart = "${xtruder.code-server}/bin/code-server";
      WorkingDirectory = "/home/kamal";
      NoNewPrivileges = true;
      User = "kamal";
      Group = "nogroup";
    };
  };
}
```

offlinehacker commented:
@wucke13 this must be issue with `code-server` itself?

wucke13 commented:
Might be the case, yes

offlinehacker commented:
@GrahamcOfBorg build code-server

offlinehacker commented:
@GrahamcOfBorg build code-server

offlinehacker commented:
One thing i noticed `minify` step is memory intensive. I tried to disable it, but had more issues, so i will leave it on.

offlinehacker commented:
@GrahamcOfBorg build code-server

offlinehacker commented:
I implemented all suggestions, improved syntax, removed unnecessary dependencies and improved comments.

Can you please check for last time if you see something that needs changing, or I think this is now really read to be merged.  @lilyball @FRidh

Mic92 commented on pkgs/servers/code-server/default.nix:
```diff
@@ -0,0 +1,187 @@
+    homepage = "https://github.com/cdr/code-server";
+    license = licenses.mit;
+    maintainers = with maintainers; [ offline ];
+    platforms = ["x86_64-linux"];
```
Since this is marked as x86_64 we can remove the aarch64 checksum above.
